### PR TITLE
Fix state when _LIBCONTAINER in environment

### PR DESCRIPTION
**Pull request by hqhq**
Fixes: #1311

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

### CODE REVIEW

**Comment by crosbymichael** [comment]
Looks like this messes with C/R 

`not ok 3 checkpoint and restore`

**Comment by hqhq** [comment]
@crosbymichael Yeah it's because restore always recreate that fifo, I pushed a commit to clean that up, ideally restore should've never created that in the first place if it won't use it, but add a new parameter to `factory.Create` seems overdo, not sure if there are other ways.

File: restore.go [hunk 1]
```diff
@@ -137,7 +138,10 @@ func restoreContainer(context *cli.Context, spec *specs.Spec, config *configs.Co
 		if err != nil {
 			return -1, err
 		}
+		// TODO: restore should not create this fifo in the first place.
+		cleanupExecFifo(context)
```
Comment (cyphar) [inline restore.go]: Yeah, I agree this a bit ugly. I think the problem boils down to the fact that `libcontainer.Factory.Create` is being used as both a "real" create (for `runc create`) but also as a "fake" create (in `runc restore`) in order to effectively make it look like `.Load` worked.

Maybe there's a nicer way of handling this? Should we move the stateful stuff out of `Create` into something separate? I'm not sure.

**Comment by hqhq** [comment]
@crosbymichael @cyphar I've updated this PR, move exec fifo's creation out of factory.Create, PTAL.

File: libcontainer/container_linux.go [hunk 2]
```diff
@@ -191,17 +191,28 @@ func (c *linuxContainer) Start(process *Process) error {
...
+	}
+	return nil
 }
 
 func (c *linuxContainer) Run(process *Process) error {
 	c.m.Lock()
-	defer c.m.Unlock()
 	status, err := c.currentStatus()
 	if err != nil {
 		return err
```
Comment (crosbymichael) [inline libcontainer/container_linux.go]: Need an unlock here

File: libcontainer/container_linux.go [hunk 3]
```diff
@@ -291,6 +302,37 @@ func (c *linuxContainer) Signal(s os.Signal, all bool) error {
...
+	if err != nil {
+		return err
+	}
+	rootgid, err := c.Config().HostGID()
+	if err != nil {
+		return err
+	}
+
+	fifoName := filepath.Join(c.root, execFifoFilename)
+	if _, err := os.Stat(fifoName); err == nil {
```
Comment (crosbymichael) [inline libcontainer/container_linux.go]: For this, should we just call mkfifo and handle an eexist error?   Does it generate an eexist error ?

Comment (hqhq) [inline libcontainer/container_linux.go]: I don't think so, `syscall.Mkfifo` will call `mknodat` syscall, and according to https://linux.die.net/man/2/mknodat ,there is no EEXIST error.

Comment (crosbymichael) [inline libcontainer/container_linux.go]: ok, SGTM

File: libcontainer/container_linux.go (hunk 2 above)
Comment (hqhq) [inline libcontainer/container_linux.go]: Updated.

**Comment by crosbymichael** [comment]
LGTM

File: libcontainer/container_linux.go [hunk 4]
```diff
@@ -291,6 +303,37 @@ func (c *linuxContainer) Signal(s os.Signal, all bool) error {
...
+		return fmt.Errorf("exec fifo %s already exists", fifoName)
+	}
+	oldMask := syscall.Umask(0000)
+	if err := syscall.Mkfifo(fifoName, 0622); err != nil {
+		syscall.Umask(oldMask)
+		return err
+	}
+	syscall.Umask(oldMask)
```
Comment (cyphar) [inline libcontainer/container_linux.go]: This is an aside, but is there a reason that we don't do `syscall.Chmod` manually? I don't really trust `umask(2)` when it comes to Go programs (where you can have goroutines run code in your current thread in-between different statements).

**Comment by cyphar** [comment]
LGTM. We can clean up the `umask` stuff later.

**Pull request merged**